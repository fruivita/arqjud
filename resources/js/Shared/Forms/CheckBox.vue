<!--
    Input to tipo checkbox com texto clicÃ¡vel.

    @see https://vuejs.org/guide/introduction.html
    @see https://tailwindcss.com/docs
    @see https://inertiajs.com/
 -->

<script>
export default { inheritAttrs: false };
</script>

<script setup>
import { gerarID } from '@/Composables/UseGerarID';
import { computed } from 'vue';

const props = defineProps({
    id: { type: String },
    label: { type: String, required: true },
    disabled: { type: Boolean, default: false },
    modelValue: { type: Boolean, required: true },
});

const emit = defineEmits(['update:modelValue']);

const idReal = computed(() => props.id ?? gerarID());

const onClick = (event) => {
    if (!props.disabled) {
        emit('update:modelValue', event.target.checked);
    }
};
</script>

<template>
    <label :for="idReal" class="inline-flex select-none items-center space-x-2" dusk="label">
        <input
            :id="idReal"
            :checked="modelValue"
            :disabled="disabled"
            @click="onClick"
            v-bind="$attrs"
            class="h-5 w-5 accent-primaria-500"
            type="checkbox"
        />

        <span>{{ label }}</span>
    </label>
</template>
