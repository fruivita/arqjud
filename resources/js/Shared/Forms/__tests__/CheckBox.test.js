/**
 * Testes para o componente CheckBox.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import CheckBox from '@/Shared/Forms/CheckBox.vue';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test, vi } from 'vitest';

vi.mock('@/Composables/UseGerarID', () => ({
    __esModule: true,
    ...vi.importActual('@/Composables/UseGerarID'),
    gerarID: () => 'foo-id',
}));

let mountFunction;

beforeEach(() => {
    mountFunction = (options = {}) => {
        return mount(CheckBox, { ...options });
    };
});

describe('CheckBox', () => {
    test('propriedades estão definidas', () => {
        expect(CheckBox.props).toMatchObject({
            id: { type: String },
            label: { type: String, required: true },
            disabled: { type: Boolean, default: false },
            modelValue: { type: Boolean, required: true },
        });
    });

    test('renderiza o componente respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { label: 'foo', modelValue: false },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o componente desabilitado respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { label: 'foo', disabled: true, modelValue: false },
            }).html()
        ).toMatchSnapshot();
    });

    test('atributos do componente e extras renderizados corretamente respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    id: '123',
                    label: 'foo',
                    modelValue: true,
                    extra: 'loren',
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('atualiza o v-model ao alterar o valor do elemento', async () => {
        const parent = mount({
            data: () => {
                return { foo: false };
            },
            template: '<div> <CheckBox v-model="foo" label="foo" /> </div>',
            components: { CheckBox: CheckBox },
        });

        const input = parent.find('input');

        await input.trigger('click');

        expect(parent.vm.foo).toBe(true);
    });

    test('dispara o evento onClick se o componente estiver habilitado', () => {
        const wrapper = mountFunction({
            props: { label: 'foo', disabled: false, modelValue: false },
        });

        const event = { target: { checked: true } };

        wrapper.vm.onClick(event);

        expect(wrapper.emitted()['update:modelValue'][0]).toEqual([true]);
    });

    test('não dispara o evento onClick se o componente estiver desabilitado', () => {
        const wrapper = mountFunction({
            props: { label: 'foo', disabled: true, modelValue: false },
        });

        const event = { target: { checked: true } };

        wrapper.vm.onClick(event);

        expect(wrapper.emitted()['update:modelValue']).toBeFalsy();
    });
});
