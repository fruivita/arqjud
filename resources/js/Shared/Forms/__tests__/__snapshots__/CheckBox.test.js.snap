// Vitest Snapshot v1

exports[`CheckBox > atributos do componente e extras renderizados corretamente respeitando o snapshot 1`] = `"<label for=\\"123\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"123\\" extra=\\"loren\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>foo</span></label>"`;

exports[`CheckBox > renderiza o componente desabilitado respeitando o snapshot 1`] = `"<label for=\\"foo-id\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"foo-id\\" disabled=\\"\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>foo</span></label>"`;

exports[`CheckBox > renderiza o componente respeitando o snapshot 1`] = `"<label for=\\"foo-id\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"foo-id\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>foo</span></label>"`;
